### ISSUE TITLE
Player

### ISSUE THREAD

**Issue by Diggsey** [GSD6338/XmasTree#9]
I made a player for the CSV files:
![image](https://user-images.githubusercontent.com/451321/147515363-ffe75933-f06c-48b2-85d2-9d2dd0e648f9.png)

https://github.com/Diggsey/xmas_tree

You just need Rust.

```
cargo run --bin xmas_tree_player --release -- sequences/barber_pole.csv
```

**Comment by Aonodensetsu** [comment]
That's a third visualizer, I guess people really like making those (I made one in Python and there has been an openGL one mentioned in #3)
Good work :)

**Comment by programer19** [comment]
NICE! About the cone in the middle... does it not block some lights? Or it is semitransparent?

**Comment by gentlegiantJGC** [comment]
I too am working on one that mirrors the python neopixel library so it can visualise any python code that could run on the real tree as well as run.py

As an added bonus it can also be used to generate the CSV animation file from the data that the python code gave it.

**Comment by Diggsey** [comment]
@programer19 currently the lights are rendered in a separate pass that uses its own depth buffer, so they are never obscured by the cone even when it's in front, mimicking the effect of seeing the lights "through" the leaves.

The lights themselves are two colocated spheres of different sizes. The inner sphere is opaque and has a +0.5 lightness modifier on the base colour. The outer sphere is semitransparent - the opacity is set to half the lightness of the base colour, so when the light is off, it's completely transparent, and when the light is bright white, it's 50% transparent. The saturation of the outer sphere is also boosted by a factor of 2 so that colours show through more intensely in the halo, as is the case in video of the tree.

**Comment by Aonodensetsu** [comment]
my visualizer can also generate csv but i'm not mimicking neopixel (although you return the values as a list of length == amount of LEDs as [r,g,b] so if i change that inner list to a tuple it's basically the same thing)

**Comment by Diggsey** [comment]
Yeah, there's a separate binary in the same repo `xmas_tree_gen` which can generate CSVs with various effects, based on a set of input coordinates.

**Comment by programer19** [comment]
the challenge of the best tree animation has become a challenge of the best tree visualization xD
I approve, this task is really interesting and worth spending time on

**Comment by Aonodensetsu** [comment]
my visualizer has semi-transparent wires ( ͡° ͜ʖ ͡°)

**Comment by Aonodensetsu** [comment]
my visualizer has semi-transparent wires ( ͡° ͜ʖ ͡°)
![image](https://user-images.githubusercontent.com/58662350/147578864-0161fcd8-0323-4a17-b6b3-261367238869.png)
and shows axes in the correct scale which might be useful when writing effects

**Comment by programer19** [comment]
Yep... much better now!
![the_best_tree_ever](https://user-images.githubusercontent.com/2873294/147613044-0fbbe726-624a-4536-a7ab-ca2989599a2f.png)

**Comment by leukipp** [comment]
[Another one](https://github.com/standupmaths/xmastree2021/pull/32) using WebGL  and WASM (#10) for online python development:
![app](https://github.com/leukipp/xmastree/raw/develop/jupyter/img/app.gif)
